src/app/profile/profile.component.ts

```diff
@@ -0,0 +1,108 @@
+import { Component, OnInit } from '@angular/core';
+import { FormGroup, FormControl, FormArray, Validators } from '@angular/forms';
+import { customEmailValidator } from './validators/email.validator';
+
+import { Profile } from './profile.model';
+import { ProfileService } from '../profile.service';
+
+@Component({
+  selector: 'app-profile',
+  templateUrl: './profile.component.html',
+  styleUrls: ['./profile.component.css']
+})
+
+export class ProfileComponent implements OnInit {
+
+  profiles: any;
+
+  profileForm!: FormGroup;
+
+  firstName = new FormControl('', Validators.required);
+  lastName = new FormControl('', Validators.required);
+  email = new FormControl('', [ Validators.required, customEmailValidator() ]);
```

Comment:
COMMENT: I think i had to set the properties here to be able to access them in the template. I saw a lot of different techniques here. Is there a specific way that is preferred?

Reply:
My preference is using the FormBuilder approach.  It's a bit more declarative, less verbose and will be better for building complex forms.

If you were to remove those properties and do something like this:

this.profileForm = new FormGroup({
      firstName: new FormControl('', Validators.required),
      lastName: new FormControl('', Validators.required),
      email: new FormControl('', [ Validators.required, customEmailValidator() ]),
      phoneNumbers: new FormArray([])
    });

You can add a helper function in your component like this to make a reusable error checker without depending on the form controls as properties:

controlHasError(controlName: string, errorType: string){
    const control = this.profileForm.get(controlName);
    return control?.getError(errorType) && control.dirty;
  }

Then in your template:
<span class="error-message" *ngIf="controlHasError('firstName', 'required')">Required Field</span>

I have a new task for you -- Why don't you go look up FormBuilder and give it a whirl? Resume wizard will use that approach.

Reply:
Got it! As I was building this I was wondering if it would have been better to use FormBuilder. Since i started without it, i just kept going with it which I think complicated things when I got to the FormArray. 

I refactored to use FormBuilder and reworked my validation errors in this commit https://github.com/lmahoney1218/angular-tour-of-heroes/pull/1/commits/3ba27e2428a7c766db34ddf9f70190e896bcee75